common/djangoapps/student/roles.py

```diff
@@ -60,21 +62,47 @@ def strict_role_checking():
     ACCESS_ROLES_INHERITANCE.update(OLD_ACCESS_ROLES_INHERITANCE)
 
 
+def get_role_cache_course_key(course_id=None):
+    """
+    Get the cache key for the course key.
+    """
+    default = ROLE_CACHE_UNGROUPED_COURSES_KEY
+    if not course_id:
+        return default
+    if not getattr(course_id, 'html_id', False):
+        return default
+    html_id = course_id.html_id()
+    if not html_id:
+        return default
+
+    return html_id
+
+
 class BulkRoleCache:  # lint-amnesty, pylint: disable=missing-class-docstring
     CACHE_NAMESPACE = "student.roles.BulkRoleCache"
     CACHE_KEY = 'roles_by_user'
 
     @classmethod
     def prefetch(cls, users):  # lint-amnesty, pylint: disable=missing-function-docstring
-        roles_by_user = defaultdict(set)
+        roles_by_user = {}
```

Comment:
I put this in the wrong place before.

> I think this might be because he's using it to mix two different key types (which I also think is worth examining). That could be done with a defaultdict, but it'd need a lambda to switch by type or something, which is getting into 😦 territory.

Reply:
done